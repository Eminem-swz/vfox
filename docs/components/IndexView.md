# IndexView/IndexViewItem 索引栏

## Import

```
import { IndexView, IndexViewItem } from 'vfox'
```

具体的引入方式可以参考[引入组件](../index.md#引入组件)。

## IndexView Props

| 属性              | 类型          | 默认值 | 必填 | 说明                        |
| ----------------- | ------------- | ------ | ---- | --------------------------- |
| sticky-offset-top | number/string | 0      | 否   | 数值默认是 px，也支持 vw/vh |

## IndexView Events

| 事件   | 描述       | 回调函数参数               |
| ------ | ---------- | -------------------------- |
| change | 切换时触发 | { activeIndex } 当前项索引 |

## Methods

| 方法名         | 说明                                                    | 参数                         |
| -------------- | ------------------------------------------------------- | ---------------------------- |
| scrollToIndex  | 将位于指定位置的 `IndexViewItem` 滚动到可视区的指定位置 | ({ index: number }) => void  |
| scrollToOffset | 滚动列表到指定的偏移，单位 px                           | ({ offset: number }) => void |

## IndexView Slots

注：其中只可放置 [IndexViewItem](./IndexView.md#indexviewitem-索引子项) 组件，否则会导致未定义的行为。

```
<fx-index-view>
  <fx-index-view-item name="A">
    <fx-cell label="单元格"></fx-cell>
    <fx-cell label="单元格"></fx-cell>
    <fx-cell label="单元格"></fx-cell>
  </fx-index-view-item>
  ...
</fx-index-view>
```

## IndexViewItem Props

| 属性 | 类型   | 默认值 | 必填 | 说明                       |
| ---- | ------ | ------ | ---- | -------------------------- |
| name | string |        | 是   | 分组名，也应用于吸附和菜单 |

## IndexViewItem Slots

```
<fx-index-view-item name="A">
  <fx-cell label="单元格"></fx-cell>
  <fx-cell label="单元格"></fx-cell>
  <fx-cell label="单元格"></fx-cell>
  ...
</fx-index-view-item>
```
