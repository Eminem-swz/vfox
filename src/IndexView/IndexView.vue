<template>
  <div :class="[prefix + '-index-view']">
    <div :class="[prefix + '-index-view_sidebar']">
      <ul :class="[prefix + '-index-view_list']" ref="list">
        <li
          :class="{ active: item.value === activeIndex }"
          v-for="item in indexList"
          :data-value="item.value"
          :key="item.value"
        >
          {{ item.label }}
        </li>
      </ul>
    </div>
    <div :class="[prefix + '-index-view_body']">
      <sticky-view
        :offset-top="stickyOffsetTop"
        ref="body"
        :activeIndex.sync="activeIndex"
        @reset-items="onResetItems"
        @change="onChange"
      >
        <slot></slot>
      </sticky-view>
    </div>
  </div>
</template>

<script>
import StickyView from '../Sticky/View.vue'
import { SDKKey } from '../config'
import { querySelector } from '../helpers/dom'
import { sizeValidator } from '../helpers/validator'
import { touchEvent } from '../helpers/events'
import { rangeInteger } from '../helpers/util'

const { touchstart, touchmove, touchend, addListeners, removeListeners, getTouch } = touchEvent

export default {
  name: SDKKey + '-index-view',
  components: { StickyView },
  props: {
    stickyOffsetTop: {
      validator: sizeValidator,
      default: 0
    }
  },
  data() {
    return {
      prefix: SDKKey,

      activeIndex: 0,

      indexList: []
    }
  },
  mounted() {
    this.resetContainer(document)

    addListeners(this.$refs.list, this)
  },
  destroyed() {},
  beforeDestroy() {
    removeListeners(this.$refs.list, this)
  },
  methods: {
    /**
     * 事件
     * @param {Event} e
     */
    handleEvent(e) {
      switch (e.type) {
        case touchstart:
          this.onTouchStart(e)
          break
        case touchmove:
          this.onTouchMove(e)
          break
        case touchend:
          this.onTouchEnd(e)
          break
        case 'mouseleave':
          this.onTouchEnd(e)
          break
        default:
          break
      }
    },

    onTouchStart(e) {
      const { clientY } = getTouch(e)

      const $target = e.target
      const value = parseInt($target.dataset.value)
      const rects = $target.getClientRects()[0]

      this.coords = {
        size: rects.height,
        offsetY: clientY - rects.top,
        startY: clientY,
        current: value
      }

      clearTimeout(this.changeTimer)
      this.changeTimer = setTimeout(() => {
        this.activeIndex = value
      }, 500)

      e.preventDefault()
    },

    onTouchMove(e) {
      if (!this.coords) {
        return
      }

      const { clientY } = getTouch(e)
      const { startY, size, offsetY, current } = this.coords

      const y = clientY - startY

      let offsetCount = 0

      if (y > 0) {
        offsetCount = Math.floor(y / size) + (y % size > size - offsetY ? 1 : 0)
      } else if (y < 0) {
        offsetCount = -Math.floor(-y / size) + (-y % size > offsetY ? -1 : 0)
      }

      if (offsetCount !== 0) {
        clearTimeout(this.changeTimer)
        this.coords.isChange = true

        this.changeTimer = this.changeTimer = setTimeout(() => {
          this.activeIndex = rangeInteger(current + offsetCount, 0, this.indexList.length - 1)
        }, 100)
      }

      e.stopPropagation()
    },

    onTouchEnd(e) {
      if (this.coords) {
        if (!this.coords.isChange) {
          this.activeIndex = this.coords.current
        }

        delete this.coords
        e.stopPropagation()
      }
    },

    resetContainer(containSelector) {
      const $container = querySelector(containSelector)

      this.$refs.body.resetContainer($container)
    },

    onResetItems(list) {
      if (this._isDestroyed) {
        return
      }

      this.indexList = list.map(({ name, index }) => {
        return {
          label: name.substr(0, 1),
          value: index
        }
      })
    },

    onChange(res) {
      this.$emit('change', res)
    },

    /**
     * 滚动到第index个
     * @param {Object} options
     */
    scrollToIndex(options) {
      this.$refs.body.scrollToIndex(options)
    },

    /**
     * 滚到到指定位置
     * @param {Object} options
     */
    scrollToOffset(options) {
      this.$refs.body.scrollToOffset(options)
    }
  }
}
</script>
