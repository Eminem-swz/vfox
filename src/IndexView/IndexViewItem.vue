<template>
  <group
    class="fx-sticky-view-item fx-index-view-item"
    :title="name"
    :data-name="name"
  >
    <slot></slot>
  </group>
</template>

<script lang="ts">
import { defineComponent, inject, onMounted, onUnmounted } from 'vue'
import Group from '@/Group'
import { createUpdateInItem } from '@/utils/list'

export default defineComponent({
  name: 'fx-index-view-item',
  components: { Group },
  props: {
    name: {
      type: String,
      required: true
    }
  },
  setup() {
    const update = inject(
      'fxStickyViewUpdate',
      createUpdateInItem('index-view')
    )

    onMounted(update)
    onUnmounted(update)

    return {}
  }
})
</script>
