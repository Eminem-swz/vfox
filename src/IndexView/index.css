.fx-index-view {
  overflow-y: auto;
  position: relative;
}

.fx-index-view_sidebar {
  position: fixed;
  right: 0;
  top: 0;
  bottom: 0;
  width: 36px;
  z-index: 1000;
  display: flex;
  flex-direction: column;
  justify-content: center;
  overflow: hidden;
}

.fx-index-view_list {
  list-style: none;
  margin: 0;
  padding: 0;
  -webkit-user-select: none;
  user-select: none;
}

.fx-index-view_list li {
  padding: 0 0 0 10px;
  text-align: center;
  height: 18px;
  line-height: 18px;
  font-size: 10px;
  color: #bfbfbf;
  font-weight: 500;
  cursor: pointer;
}

.fx-index-view_list li.active {
  color: #1890ff;
}

.fx-index-view-item {
  margin-bottom: 0;
}

.fx-index-view-item .fx-group_header {
  height: 28px;
  background-color: #f5f5f5;
}
