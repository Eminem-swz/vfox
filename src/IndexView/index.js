import IndexView from './IndexView.vue'

IndexView.install = function(Vue) {
  Vue.component(IndexView.name, IndexView)
}

export default IndexView
