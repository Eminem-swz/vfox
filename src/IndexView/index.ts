import { App } from 'vue'
import { SFCWithInstall } from '@/helpers/types'
import IndexView from './IndexView.vue'

const _IndexView: SFCWithInstall<typeof IndexView> = Object.assign(IndexView, {
  install: function (app: App) {
    app.component(IndexView.name, IndexView)
  }
})

export default _IndexView
