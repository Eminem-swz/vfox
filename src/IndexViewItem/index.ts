import { withNoopInstall } from '@/helpers/with-install'
import { IndexViewItem } from '@/IndexView'

export { IndexViewItem }
export default withNoopInstall(IndexViewItem)
