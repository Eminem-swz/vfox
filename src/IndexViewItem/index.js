import IndexViewItem from '../IndexView/IndexViewItem.vue'

IndexViewItem.install = function(app) {
  app.component(IndexViewItem.name, IndexViewItem)
}

export default IndexViewItem
