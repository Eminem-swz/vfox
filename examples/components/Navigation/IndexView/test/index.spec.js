/* global describe,test,expect */

import { mount } from '@vue/test-utils'
import IndexView from '@/IndexView'
import IndexViewItem from '@/IndexViewItem'

describe('IndexView', () => {
  test('IndexView should render default correctly', () => {
    const wrapper = mount(IndexView, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('IndexViewItem should render default correctly', () => {
    const wrapper = mount(IndexViewItem, {
      props: { name: 'item-1' },
      slots: {
        default: 'content'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
