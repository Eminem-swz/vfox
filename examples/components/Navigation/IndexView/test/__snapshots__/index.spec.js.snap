// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexView IndexView should render default correctly 1`] = `
<div class="fx-index-view">
  <div class="fx-index-view_sidebar">
    <ul class="fx-index-view_list"></ul>
  </div>
  <div class="fx-index-view_body">
    <div class="fx-sticky-view">
      <div class="fx-sticky-view_list"></div>
      <div class="fx-sticky fx-sticky-view_top">
        <div class="fx-sticky_content" style="top: 0px; left: 0px; width: 0px; height: 0px; z-index: 1000; position: fixed;">
          <div class="fx-sticky-view_fixed" style="transform: translate3d(0px, 0px, 0px);"></div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`IndexView IndexViewItem should render default correctly 1`] = `
<div class="fx-group fx-horizontal-hairline fx-sticky-view-item fx-index-view-item">
  <div class="fx-group_header">
    <div class="fx-group_title">item-1</div>
    <div class="fx-group_more"></div>
  </div>
  <div class="fx-group_body fx-horizontal-hairline hairline-reverse">
    <div class="fx-group_body-inner">content</div>
  </div>
</div>
`;
